src/vos/vos_internal.h

```diff
@@ -332,9 +332,56 @@ struct vos_dtx_cmt_ent {
 #define DCE_EPOCH(dce)		((dce)->dce_base.dce_epoch)
 #define DCE_HANDLE_TIME(dce)	((dce)->dce_base.dce_handle_time)
 
-extern int vos_evt_feats;
+extern uint64_t vos_evt_feats;
 
 #define VOS_KEY_CMP_LEXICAL	(1ULL << 63)
+#define VOS_TREE_AGG_OPT	EVT_FEAT_AGG_OPT
+#define VOS_TREE_AGG_HLC	EVT_FEAT_AGG_HLC
+#define VOS_TREE_AGG_TIME_MASK	EVT_FEAT_AGG_TIME_MASK
+
+#define CHECK_VOS_TREE_FLAG(flag)	\
+	D_CASSERT(((flag) & (EVT_FEATS_SUPPORTED | BTR_FEAT_MASK)) == 0)
+CHECK_VOS_TREE_FLAG(VOS_KEY_CMP_LEXICAL);
+CHECK_VOS_TREE_FLAG(VOS_TREE_AGG_OPT);
+CHECK_VOS_TREE_FLAG(VOS_TREE_AGG_TIME_MASK);
+
+/** Update the aggregatable write timestamp within 1/4 second granularity */
+static inline void
+vos_feats_agg_time_update(daos_epoch_t epoch, uint64_t *feats)
+{
+	uint64_t extra_flag = 0;
+
+	if (epoch & 0xffffffff00000000ULL) {
+		/** Assume aggregation only happens at most once every quarter second */
+		epoch &= 0xffffffff00000000ULL;
+		/** Ensure the resulting epoch is not zero regardless, mostly for standalone */
+		epoch = (epoch >> 3);
+		extra_flag = VOS_TREE_AGG_HLC;
+	} else {
+		epoch &= 0x0ffffffffULL;
+		epoch = (epoch << 29);
```

Comment:
Is this for the integer epoch used in tests (or some other standalone VOS cases)? then I'm wondering how to compare it with HAE which is represented by HLC.

I'm wondering if we should always set the agg_time as maximum value in such case? so that the VOS_TREE_AGG_HLC bit could be saved?

Reply:
Oh, never mind, I forgot that HAE is from epoch but not HLC.